### PR TITLE
[P2P][Core] P2P Retries + Debug CLI Improvements

### PR DESCRIPTION
**Pull request by Olshansk**
## Description

Revive docker-compose LocalNet by fixing a few issues on the client CLI, adding retries in P2P, while also making improvements to the debug CLI.

## Issue

NA

## Type of change

Please mark the relevant option(s):

- [ ] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [ ] Other 

## List of changes

- Update makefile so the CLI targets work
- Split the debug commands into one that has a UI and one that is non-interactive
- Add PrintNodeState and TriggerNextView to the debug command w/o a UI
- Update a couple techdebt comments
- Make p2p host connections retriable with a constant backoff

## Testing

- [x] `make develop_test`; if any code changes were made
- [x] `make test_e2e` on [k8s LocalNet](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md); if any code changes were made
- [ ] `e2e-devnet-test` passes tests on [DevNet](https://pocketnetwork.notion.site/How-to-DevNet-ff1598f27efe44c09f34e2aa0051f0dd); if any code was changed
- [x] [Docker Compose LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md); if any major functionality was changed or introduced
- [x] [k8s LocalNet](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md); if any infrastructure or configuration changes were made

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added, or updated, [`godoc` format comments](https://go.dev/blog/godoc) on touched members (see: [tip.golang.org/doc/comment](https://tip.golang.org/doc/comment))
- [x] I have tested my changes using the available tooling
- [ ] I have updated the corresponding CHANGELOG

### If Applicable Checklist

- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

### CODE REVIEW

File: e2e/tests/steps_init_test.go [hunk 1]
```diff
@@ -161,6 +160,8 @@ func (s *rootSuite) getPrivateKey(
 // getClientset uses the default path `$HOME/.kube/config` to build a kubeconfig
 // and then connects to that cluster and returns a *Clientset or an error
 func getClientset(t gocuke.TestingT) (*kubernetes.Clientset, error) {
+	t.Helper()
```
Comment (dylanlott) [inline e2e/tests/steps_init_test.go:163]: 💯

File: p2p/background/router.go [hunk 2]
```diff
@@ -342,13 +349,30 @@ func (rtr *backgroundRouter) bootstrap(ctx context.Context) error {
...
+		if err := rtr.connectWithRetry(ctx, libp2pAddrInfo); err != nil {
 			return fmt.Errorf("connecting to peer: %w", err)
 		}
 	}
 	return nil
 }
 
+// connectWithRetry attempts to connect to the given peer, retrying up to connectMaxRetries times
+// and waiting connectRetryTimeout between each attempt.
+func (rtr *backgroundRouter) connectWithRetry(ctx context.Context, libp2pAddrInfo libp2pPeer.AddrInfo) error {
```
Comment (dylanlott) [inline p2p/background/router.go:361]: Nit: Is it worth considering exponential back-off here?

File: p2p/background/router.go [hunk 3]
```diff
@@ -32,6 +33,12 @@ var (
 	_ backgroundRouterFactory  = &backgroundRouter{}
 )
 
+// TECHDEBT: Make these values configurable
```
Comment (Olshansk) [inline p2p/background/router.go:36]:
```suggestion
// TECHDEBT: Make these values configurable
// TECHDEBT: Consider using an exponential backoff instead
```

File: p2p/background/router.go (hunk 2 above)
Comment (Olshansk) [reply p2p/background/router.go:361]: https://github.com/pokt-network/pocket/pull/902/files/37b2e5cd4a0035fd64737ec01be36b0170dcd903#r1261840776

File: app/client/cli/debug.go [hunk 4]
```diff
@@ -66,17 +70,65 @@ func NewDebugSubCommands() []*cobra.Command {
...
 }
 
-// NewDebugCommand returns the cobra CLI for the Debug command.
-func NewDebugCommand() *cobra.Command {
+// newDebugUICommand returns the cobra CLI for the Debug UI interface.
+func newDebugUICommand() *cobra.Command {
 	return &cobra.Command{
-		Use:               "debug",
-		Short:             "Debug utility for rapid development",
+		Use:               "debug_ui",
```
Comment (dylanlott) [inline app/client/cli/debug.go:76]: Nit: I'm generally not a fan of underscores in CLI command names. Consider removing the underscore. Sidenote: This command is used often enough it is probably worth adding an alias to `d` or something.

Comment (Olshansk) [reply app/client/cli/debug.go]: [`43ef141` (#902)](https://github.com/pokt-network/pocket/pull/902/commits/43ef141122b4e7385702596f068d4dc40372cd2b)

**Review by dylanlott — APPROVED** [review]
💯 LGTM

File: app/client/cli/debug.go (hunk 4 above)
Comment (dylanlott) [reply app/client/cli/debug.go]: Lol @ `dui`. It's perfect, let's never change it.

File: Makefile [hunk 5]
```diff
@@ -159,7 +159,7 @@ rebuild_client_start: docker_check ## Rebuild and run a client daemon which is o
 
 .PHONY: client_connect
 client_connect: docker_check ## Connect to the running client debugging daemon
-	docker exec -it client /bin/bash -c "POCKET_P2P_IS_CLIENT_ONLY=true go run -tags=debug app/client/*.go debug --remote_cli_url=http://validator1:50832"
+	docker exec -it client /bin/bash -c "POCKET_REMOTE_CLI_URL=http://validator1:50832 go run -tags=debug app/client/*.go debugui"
```
Comment (bryanchriswhite) [inline Makefile:162]: :100:

File: Makefile [hunk 6]
```diff
@@ -522,11 +522,11 @@ localnet_up: ## Starts up a k8s LocalNet with all necessary dependencies (tl;dr
 
 .PHONY: localnet_client_debug
 localnet_client_debug: ## Opens a `client debug` cli to interact with blockchain (e.g. change pacemaker mode, reset to genesis, etc). Though the node binary updates automatiacally on every code change (i.e. hot reloads), if client is already open you need to re-run this command to execute freshly compiled binary.
-	kubectl exec -it deploy/dev-cli-client --container pocket -- p1 debug --remote_cli_url http://pocket-validators:50832
+	kubectl exec -it deploy/dev-cli-client --container pocket -- p1 debugui --remote_cli_url http://pocket-validators:50832
 
 .PHONY: localnet_shell
 localnet_shell: ## Opens a shell in the pod that has the `client` cli available. The binary updates automatically whenever the code changes (i.e. hot reloads).
-	kubectl exec -it deploy/dev-cli-client --container pocket -- /bin/bash
+	kubectl exec -it deploy/dev-cli-client --container pocket -- /bin/bash -c "export POCKET_REMOTE_CLI_URL=http://pocket-validators:50832; bash"
```
Comment (bryanchriswhite) [inline Makefile:529]: Similar question / observation to the above. Is there a reason not to put this in the [cli-client mainfest](https://github.com/pokt-network/pocket/blob/main/build/localnet/manifests/cli-client.yaml#L77) instead?

Again, it looks like it's already set there (to a different value). Additionally, if we're removing `POCKET_P2P_IS_CLIENT_ONLY` we should remove it in [the manifest](https://github.com/pokt-network/pocket/blob/main/build/localnet/manifests/cli-client.yaml#L41) as well.

File: Makefile (hunk 5 above)
Comment (bryanchriswhite) [reply Makefile:162]: Can you elaborate on why remove `POCKET_P2P_IS_CLIENT_ONLY`?

File: p2p/background/router.go [hunk 7]
```diff
@@ -342,13 +350,30 @@ func (rtr *backgroundRouter) bootstrap(ctx context.Context) error {
 			return nil
 		}
 
-		if err := rtr.host.Connect(ctx, libp2pAddrInfo); err != nil {
+		if err := rtr.connectWithRetry(ctx, libp2pAddrInfo); err != nil {
```
Comment (bryanchriswhite) [inline p2p/background/router.go:353]: Can you elaborate on the rationale behind this change?

File: Makefile [hunk 8]
```diff
@@ -522,11 +522,11 @@ localnet_up: ## Starts up a k8s LocalNet with all necessary dependencies (tl;dr
 
 .PHONY: localnet_client_debug
 localnet_client_debug: ## Opens a `client debug` cli to interact with blockchain (e.g. change pacemaker mode, reset to genesis, etc). Though the node binary updates automatiacally on every code change (i.e. hot reloads), if client is already open you need to re-run this command to execute freshly compiled binary.
-	kubectl exec -it deploy/dev-cli-client --container pocket -- p1 debug --remote_cli_url http://pocket-validators:50832
+	kubectl exec -it deploy/dev-cli-client --container pocket -- p1 debugui --remote_cli_url http://pocket-validators:50832
```
Comment (bryanchriswhite) [inline Makefile:525]: :ok_hand:

File: app/client/cli/debug.go [hunk 9]
```diff
@@ -66,17 +70,66 @@ func NewDebugSubCommands() []*cobra.Command {
...
 }
 
-// NewDebugCommand returns the cobra CLI for the Debug command.
-func NewDebugCommand() *cobra.Command {
+// newDebugUICommand returns the cobra CLI for the Debug UI interface.
+func newDebugUICommand() *cobra.Command {
 	return &cobra.Command{
-		Use:               "debug",
-		Short:             "Debug utility for rapid development",
+		Aliases:           []string{"dui", "d"},
```
Comment (bryanchriswhite) [inline app/client/cli/debug.go:76]: Nice :sunglasses: 

_(debug responsibly)_

File: app/client/cli/debug.go [hunk 10]
```diff
@@ -66,17 +70,66 @@ func NewDebugSubCommands() []*cobra.Command {
...
+	}
+}
+
+func debugCommands() []*cobra.Command {
+	cmds := []*cobra.Command{
+		{
+			Use:     "TriggerView",
+			Short:   "Trigger the next view in consensus",
+			Long:    "Sends a message to all visible nodes on the network to start the next view (height/step/round) in consensus",
+			Aliases: []string{"triggerView"},
```
Comment (bryanchriswhite) [inline app/client/cli/debug.go:101]: :100: :pray: love this change!

Is it worth considering making the PascalCase the "canonical" form (in `Use`) and moving the dromedaryCase form to be the alias?

My thinking is that the dromedary form is more conventional, in general.

File: app/client/cli/helpers/setup.go [hunk 11]
```diff
@@ -18,6 +20,11 @@ func P2PDependenciesPreRunE(cmd *cobra.Command, _ []string) error {
 	// TECHDEBT: this is to keep backwards compatibility with localnet
 	flags.ConfigPath = runtime.GetEnv("CONFIG_PATH", "build/config/config.validator1.json")
```
Comment (bryanchriswhite) [inline app/client/cli/helpers/setup.go:21]: @dylanlott do you recall the rationale behind this line. It doesn't look right to me - is this still correct?

File: app/client/cli/helpers/setup.go [hunk 12]
```diff
@@ -18,6 +20,11 @@ func P2PDependenciesPreRunE(cmd *cobra.Command, _ []string) error {
...
 
+	// by this time, the config path should be set
+	configs.ParseConfig(flags.ConfigPath)
```
Comment (bryanchriswhite) [inline app/client/cli/helpers/setup.go:24]: Is this call just for side effects in viper?

I would've expected us to either do something with the config object that it returns (e.g. pass it to the runtime manager) OR use the runtime manager to load the same config file.

File: app/client/cli/helpers/setup.go (hunk 11 above)
Comment (bryanchriswhite) [reply app/client/cli/helpers/setup.go:21]: When I looked into it I got the impression that that file wasn't present in k8s localnet and as a result the default config + env vars was being used.

File: e2e/tests/steps_init_test.go [hunk 13]
```diff
@@ -23,8 +23,8 @@ import (
 var e2eLogger = pocketLogger.Global.CreateLoggerForModule("e2e")
 
 const (
-	// defines the host & port scheme that LocalNet uses for naming validators.
-	// e.g. validator-001 thru validator-999
+	// Each actor is represented e.g. validator-001 thru validator-999
+	// Defines the host & port scheme that LocalNet uses for naming actors.
 	validatorServiceURLTmpl = "validator-%s-pocket:%d"
```
Comment (bryanchriswhite) [inline e2e/tests/steps_init_test.go:28]: The comment doesn't seem to match the pattern it's describing.

```suggestion
	// Each actor is represented e.g. validator-001-pocket:42069 thru validator-999-pocket:42069.
	// Defines the host & port scheme that LocalNet uses for naming actors.
	validatorServiceURLTmpl = "validator-%s-pocket:%d"
```

**Review by bryanchriswhite — CHANGES_REQUESTED** [review]
Nice improvements! :raised_hands: 

I had some questions and suggestions but those notwithstanding, I'm confident that the CLI appreciates the love. :heart:

File: app/client/cli/debug.go (hunk 4 above)
Comment (Olshansk) [reply app/client/cli/debug.go]: Just`dui`t

File: Makefile (hunk 5 above)
Comment (Olshansk) [reply Makefile:162]: > Is there a reason not to move this into the [docker-compose.yaml](https://github.com/pokt-network/pocket/blob/main/build/deployments/docker-compose.yaml?rgh-link-date=2023-07-13T06%3A39%3A54Z#L24).

Nope.

> It looks like it's already set there, wdyt about just removing it here and relying on the env var in docker-compose.yaml?

I prefer this as well. Will give it a shot and report

> Can you elaborate on why remove POCKET_P2P_IS_CLIENT_ONLY?

Unless you have reason not to, I SHOULD remove it from `P2PConfig` in this PR as well. I do not believe it's being used anywhere.

Given that we have this function which is unrelated:

```go
func (m *p2pModule) isClientDebugMode() bool {
	return m.GetBus().GetRuntimeMgr().GetConfig().ClientDebugMode
}
```

File: p2p/background/router.go (hunk 7 above)
Comment (Olshansk) [reply p2p/background/router.go:353]: 1. #892 is a 500 line PR that I haven't had a chance to dig into yet and I'm not 100% sure we'll be able to get it in by end-of-week

2. I briefly looked at https://github.com/pokt-network/pocket/commit/1f05989cf6022a6d1fdb709a00430bea961859f7 but need a bit more time to think/digest.

3. I believe the two changes are still complementary and `connectWithRetry` is a fast (and logical) way to go around the race condition, which personally unblocked my development as well.

Lmk wyt

File: app/client/cli/helpers/setup.go (hunk 12 above)
Comment (Olshansk) [reply app/client/cli/helpers/setup.go:26]: > Is this call just for side effects in viper?

Yes. I unfortunately do not remember the exact reasoning but it was needed in relation to the changes I made in the Makefile

File: Makefile (hunk 5 above)
Comment (bryanchriswhite) [reply Makefile]: That helper was just there to support the existing functionality but I don't think that config is a dependency of any necessary logic.

File: app/client/cli/helpers/setup.go (hunk 12 above)
Comment (bryanchriswhite) [reply app/client/cli/helpers/setup.go]: (see: https://github.com/pokt-network/pocket/pull/891#discussion_r1263340347)

File: Makefile [hunk 14]
```diff
@@ -525,7 +525,7 @@ localnet_up: ## Starts up a k8s LocalNet with all necessary dependencies (tl;dr
 
 .PHONY: localnet_client_debug
 localnet_client_debug: ## Opens a `client debug` cli to interact with blockchain (e.g. change pacemaker mode, reset to genesis, etc). Though the node binary updates automatiacally on every code change (i.e. hot reloads), if client is already open you need to re-run this command to execute freshly compiled binary.
-	kubectl exec -it deploy/dev-cli-client --container pocket -- p1 debug --remote_cli_url http://pocket-validators:50832
+	kubectl exec -it deploy/dev-cli-client --container pocket -- p1 DebugUI --remote_cli_url http://pocket-validators:50832
```
Comment (bryanchriswhite) [inline Makefile:528]: Is there a reason that we're using a different remote CLI URL here and/or should we prefer the ENV var for this case?

**Review by bryanchriswhite — APPROVED** [review]
I left a couple more comments but otherwise approving, I trust your judgement :+1: :sweat_smile:

File: app/client/cli/helpers/setup.go (hunk 12 above)
Comment (Olshansk) [reply app/client/cli/helpers/setup.go]: #907

**Pull request merged**